### PR TITLE
add option to force usage of fallback_auth_store

### PR DESCRIPTION
**Pull request by 0xbe7a**
When running within a CI, it is sometimes easier not to rely on the native authentication toolchain, but to use a simple credentials file instead (see https://github.com/prefix-dev/rattler-build/issues/334). This PR adds an option to force the use of this fallback file and disable the use of the system toolchain.

### CODE REVIEW

**Comment by wolfv** [comment]
I think this is "OK" for me. Not a super-fan of the option :)

I am wondering if we should rather create another authentication storage that can be supplied instead of the current one. The new one would rely only on the fallback storage and uses plain JSON as database, and then we would have a Rust trait for an authentication storage?

**Comment by 0xbe7a** [comment]
Yeah, that sounds a bit more elegant.

File: crates/rattler_networking/src/authentication_storage/storage.rs [hunk 1]
```diff
@@ -140,6 +162,10 @@ impl AuthenticationStorage {
             cache.remove(host);
         }
 
+        if self.force_fallback_storage {
```
Comment (wolfv) [inline crates/rattler_networking/src/authentication_storage/storage.rs:165]: At the end we do also delete the password from any fallback storage. So maybe this is not needed here?

**Review by wolfv — CHANGES_REQUESTED** [review]

**Comment by 0xbe7a** [comment]
I still need to test this with rattler-build

**Comment by wolfv** [comment]
Nice work! Looks good to me. If we can get rid of anyhow it's perfect :)

**Comment by 0xbe7a** [comment]
I adapted and tested the rattler-build PR with our real-world example. Works great! I also changed the file location of the `.json` file, as the format changed slightly and we now store valid JSON.

**Comment by 0xbe7a** [comment]
Would you prefer to return a `Box<dyn Error + Send + Sync>`instead of an anyhow `Result`? Since I wanted to leave the option for users to bring your own backend, you can't really type the error type.

**Comment by wolfv** [comment]
right. let's try to ask @baszalmstra what he thinks is best.

File: crates/rattler_networking/src/authentication_storage/backends/keyring.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        Ok(())
+    }
+
+    fn get(&self, host: &str) -> Result<Option<Authentication>> {
+        let entry = Entry::new(&self.store_key, host)?;
+        let password = entry.get_password();
+
+        let p_string = match password {
+            Ok(password) => password,
+            Err(keyring::Error::NoEntry) => {
```
Comment (baszalmstra) [inline crates/rattler_networking/src/authentication_storage/backends/keyring.rs:67]: Both err statements seem to donthe same.

File: crates/rattler_networking/src/authentication_storage/storage.rs [hunk 3]
```diff
@@ -1,155 +1,100 @@
...
+/// This struct implements storage and access of authentication
+/// information backed by multiple storage backends
+/// (e.g. keyring and file storage)
+/// Credentials are stored and retrieved from the backends in the
+/// order they are added to the storage
 pub struct AuthenticationStorage {
-    /// The store_key needs to be unique per program as it is stored
-    /// in a global dictionary in the operating system
-    pub store_key: String,
+    backends: Vec<Arc<Box<dyn StorageBackend + Send + Sync>>>,
```
Comment (baszalmstra) [inline crates/rattler_networking/src/authentication_storage/storage.rs:24]: Arc Box seems superfluous. Maybe just Arc is enough?

File: crates/rattler_networking/src/authentication_storage/storage.rs [hunk 4]
```diff
@@ -190,4 +135,20 @@ impl AuthenticationStorage {
...
+            cache.insert(host.to_string(), None);
+        }
+
+        for backend in &self.backends {
+            if let Err(e) = backend.delete(host) {
+                tracing::warn!("Error deleting credentials from backend: {}", e);
+            }
+        }
+
+        Ok(())
```
Comment (baszalmstra) [inline crates/rattler_networking/src/authentication_storage/storage.rs:152]: Shouldnt this return an error if it failed?

**Review by baszalmstra — COMMENTED** [review]
Thanks for this PR, almost looks good to me. I left some small comments.

And I would just use anyhow Result.

File: crates/rattler_networking/src/authentication_storage/storage.rs (hunk 3 above)
Comment (0xbe7a) [reply crates/rattler_networking/src/authentication_storage/storage.rs:24]: We want Arc for `Clone` and need to Box to store the `dyn Trait`

File: crates/rattler_networking/src/authentication_storage/storage.rs (hunk 4 above)
Comment (0xbe7a) [reply crates/rattler_networking/src/authentication_storage/storage.rs:152]: What does "fail" mean here? An attempt is made to delete the credentials in _all_ known backends, regardless of whether the previous backend returned an error.

File: crates/rattler_networking/src/authentication_storage/storage.rs (hunk 3 above)
Comment (baszalmstra) [reply crates/rattler_networking/src/authentication_storage/storage.rs:24]: but cant you use the arc for dyn too?

File: crates/rattler_networking/src/authentication_storage/storage.rs (hunk 4 above)
Comment (baszalmstra) [reply crates/rattler_networking/src/authentication_storage/storage.rs:152]: Well each individual backend might fail but this function will still report success. I think if any of them fail the function should return an error too.

File: crates/rattler_networking/src/authentication_storage/storage.rs (hunk 3 above)
Comment (0xbe7a) [reply crates/rattler_networking/src/authentication_storage/storage.rs]: You're right, I didn't know that. Thanks!

File: crates/rattler_networking/src/authentication_storage/storage.rs (hunk 4 above)
Comment (0xbe7a) [reply crates/rattler_networking/src/authentication_storage/storage.rs:152]: But then the fallback behavior differs from `store/get`, where errors are only logged and the next backend is used. What do you think about returning an error if all backends fail, which would be consistent with the other methods?

Comment (baszalmstra) [reply crates/rattler_networking/src/authentication_storage/storage.rs:152]: Yeah that sounds good! Maybe that should also be the behavior for the other methods?

**Review by baszalmstra — APPROVED** [review]

**Review by wolfv — APPROVED** [review]

**Pull request merged**